src/main/java/org/apache/sysds/runtime/instructions/fed/TernaryFEDInstruction.java

```diff
@@ -0,0 +1,160 @@
+
+	private TernaryFEDInstruction(TernaryOperator op, CPOperand in1, CPOperand in2, CPOperand in3, CPOperand out, String opcode, String str) {
+		super(FEDInstruction.FEDType.Ternary, op, in1, in2, in3, out, opcode, str);
+	}
+
+	public static TernaryFEDInstruction parseInstruction(String str)
+	{
+		String[] parts = InstructionUtils.getInstructionPartsWithValueType(str);
+		String opcode=parts[0];
+		CPOperand operand1 = new CPOperand(parts[1]);
+		CPOperand operand2 = new CPOperand(parts[2]);
+		CPOperand operand3 = new CPOperand(parts[3]);
+		CPOperand outOperand = new CPOperand(parts[4]);
+		TernaryOperator op = InstructionUtils.parseTernaryOperator(opcode);
+		return new TernaryFEDInstruction(op, operand1, operand2, operand3, outOperand, opcode, str);
+	}
+
+	@Override
+	public void processInstruction( ExecutionContext ec ) {
+		MatrixObject mo1 = input1.isMatrix() ? ec.getMatrixObject(input1.getName()) : null;
+		MatrixObject mo2 = input2.isMatrix() ? ec.getMatrixObject(input2.getName()) : null;
+		MatrixObject mo3 = input3.isMatrix() ? ec.getMatrixObject(input3.getName()) : null;
+
+		long matrixInputsCount = List.of(mo1, mo2, mo3).stream().filter(Objects::nonNull).count();
+
+		if(matrixInputsCount == 3)
+			processMatrixInput(ec, mo1, mo2, mo3);
+		else if (matrixInputsCount == 1) {
+			CPOperand in = mo1 == null ? mo2 == null ? input3 : input2 : input1;
+			mo1 = mo1 == null ? mo2 == null ? mo3 : mo2 : mo1;
+			processMatrixScalarInput(ec, mo1, in);
+		} else
+			process2MatrixScalarInput(ec, mo1, mo2, mo3);
+	}
+
+	private void processMatrixScalarInput(ExecutionContext ec, MatrixObject mo1, CPOperand in) {
+		FederatedRequest fr1 = FederationUtils.callInstruction(instString, output,
+			new CPOperand[] {in}, new long[] {mo1.getFedMapping().getID()});
+		mo1.getFedMapping().execute(getTID(), true, fr1);
+
+		//derive new fed mapping for output
+		MatrixObject out = ec.getMatrixObject(output);
+		out.getDataCharacteristics().set(mo1.getDataCharacteristics());
+		out.setFedMapping(mo1.getFedMapping().copyWithNewID(fr1.getID()));
+	}
+
+	private void process2MatrixScalarInput(ExecutionContext ec, MatrixObject mo1, MatrixObject mo2, MatrixObject mo3) {
+		CPOperand[] inputArgs = new CPOperand[] {input1, input2};
+		if(mo1 != null && mo1.isFederated() && mo2 == null) {
+			mo2 = mo3;
+			inputArgs = new CPOperand[] {input1, input3};
+		} else if(mo2 != null && mo2.isFederated() && mo1 == null) {
+			mo1 = mo2;
+			mo2 = mo3;
+			inputArgs = new CPOperand[] {input2, input3};
+		} else if(mo2 != null && mo2.isFederated() && mo1 != null) {
+			mo1 = mo2;
+			mo2 = ec.getMatrixObject(input1);
+			inputArgs = new CPOperand[] {input2, input1};
+		} else if(mo3 != null && mo3.isFederated() && mo1 == null) {
+			mo1 = mo3;
+			inputArgs = new CPOperand[] {input3, input2};
+		} else if(mo3 != null && mo3.isFederated() && mo1 != null) {
+			mo1 = mo3;
+			mo2 = ec.getMatrixObject(input1);
+
+			inputArgs = new CPOperand[] {input3, input1};
+		}
```

Comment:
Is it necessary to swap the variables like this? The functionality is fine, but it is difficult to read and to follow where the matrix objects end up by the end of the if-else statements.

Reply:
I modified it slightly, so now it's more readable.